Added MaterialDialog for visual feedback when clicking (Take Your Pictures).

jaindiv26 opened the pull request:
## Description
Added MaterialDialog for visual feedback when clicking (Take Your Pictures). Changes have been in "OpenFoodAPIClient" file.
## Related issues and discussion
#1978 There is no visual feedback on uploading image using "Take more pictures"
 
 ## Screen-shots

![screenshot_2018-11-14-23-17-53-269](https://user-images.githubusercontent.com/32436867/48501926-6396c900-e864-11e8-8b18-1f288c9b7e5b.jpeg)

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -339,12 +340,20 @@ public void onFailure(Call<State> call, Throwable t) {
         lt.show();**/
-
+        Dialog dialog;
+        MaterialDialog.Builder builder = new MaterialDialog.Builder(context)
```
No need for builder variable since it's not used elsewhere, just use dialog.

jaindiv26 replied:
Yeah, I did that changes and in the next commit @brandenfung you can see that.

teolemon commented:
I would like the dialog to be non blocking. This way, the user can take a second picture without waiting for the first one to finish uploading

jaindiv26 commented:
> I would like the dialog to be non blocking. This way, the user can take a second picture without waiting for the first one to finish uploading

I'll make the dialog cancellable to true but then the user will think that the uploading is canceled

jaindiv26 commented:
> Can the spinner be directly in the view, above the button, and smaller ?

I think it's perfect.

teolemon commented:
![image](https://user-images.githubusercontent.com/1689815/48635275-02880600-e9c8-11e8-8a95-b6f5108f9f8c.png)

jaindiv26 commented:
> In the edit mode, the spinner is already in the view.
> ![image](https://user-images.githubusercontent.com/1689815/48635275-02880600-e9c8-11e8-8a95-b6f5108f9f8c.png)

Ok @teolemon I'll make it like this.

jaindiv26 commented:
@teolemon I've added a listener that will help to get the status of the Image Upload and removed the Material Dialog.

![screenshot_2018-11-24-21-47-39-496](https://user-images.githubusercontent.com/32436867/48970483-b88bc980-f032-11e8-8e25-1f9e0a6e1f75.jpeg)
![screenshot_2018-11-24-21-47-42-772](https://user-images.githubusercontent.com/32436867/48970484-b88bc980-f032-11e8-83f2-a25bfcd3bfb7.jpeg)

teolemon commented:
@Karljoones @brandenfung The behaviour was changed.

teolemon commented:
Crash when trying to take a picture in airplane mode
```
2018-11-28 14:17:33.303 12652-12652/? E/AndroidRuntime: FATAL EXCEPTION: main
    Process: openfoodfacts.github.scrachx.openfood, PID: 12652
    java.lang.NullPointerException: Attempt to invoke virtual method 'java.lang.String android.content.Context.getPackageName()' on a null object reference
        at android.widget.Toast.<init>(Toast.java:117)
        at android.widget.Toast.makeText(Toast.java:280)
        at android.widget.Toast.makeText(Toast.java:270)
        at openfoodfacts.github.scrachx.openfood.views.product.summary.SummaryProductFragment.onFailure(SummaryProductFragment.java:826)
        at openfoodfacts.github.scrachx.openfood.network.OpenFoodAPIClient$9.onFailure(OpenFoodAPIClient.java:379)
        at retrofit2.ExecutorCallAdapterFactory$ExecutorCallbackCall$1$2.run(ExecutorCallAdapterFactory.java:79)
        at android.os.Handler.handleCallback(Handler.java:873)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:193)
        at android.app.ActivityThread.main(ActivityThread.java:6669)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:493)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:858)
```

jaindiv26 commented:
@teolemon I'll look into this crash.

jaindiv26 commented:
@teolemon I've fixed the problem, please review.

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -201,17 +203,17 @@ public void onResponseCallForPostFunction(Call<State> call, Response<State> resp
         if (StringUtils.isNotEmpty(imguploadFront)) {
             ProductImage image = new ProductImage(product.getBarcode(), FRONT, new File(imguploadFront));
-            postImg(activity, image);
+            postImg(activity, image,null);
```
```suggestion
            postImg(activity, image, null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -201,17 +203,17 @@ public void onResponseCallForPostFunction(Call<State> call, Response<State> resp
         String imguploadIngredients = product.getImgupload_ingredients();
         if (StringUtils.isNotEmpty(imguploadIngredients)) {
-            postImg(activity, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)));
+            postImg(activity, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)),null);
```
```suggestion
            postImg(activity, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)), null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -201,17 +203,17 @@ public void onResponseCallForPostFunction(Call<State> call, Response<State> resp
         String imguploadNutrition = product.getImgupload_nutrition();
         if (StringUtils.isNotBlank(imguploadNutrition)) {
-            postImg(activity, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)));
+            postImg(activity, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)),null);
```
```suggestion
            postImg(activity, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)), null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -349,12 +350,18 @@ public void onResponse(@NonNull Call<JsonNode> call, @NonNull Response<JsonNode>
                             ToUploadProduct product = new ToUploadProduct(image.getBarcode(), image.getFilePath(), image.getImageField().toString());
                             mToUploadProductDao.insertOrReplace(product);
                             Toast.makeText(context, response.toString(), Toast.LENGTH_LONG).show();
+                            if(imageUploadListener !=null){
```
```suggestion
                            if(imageUploadListener != null){
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -349,12 +350,18 @@ public void onResponse(@NonNull Call<JsonNode> call, @NonNull Response<JsonNode>
 
                         JsonNode body = response.body();
                         Log.d("onResponse", body.toString());
+                        if(imageUploadListener !=null){
```
```suggestion
                        if(imageUploadListener != null){
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -368,6 +375,9 @@ public void onResponse(@NonNull Call<JsonNode> call, @NonNull Response<JsonNode>
                     @Override
                     public void onFailure(@NonNull Call<JsonNode> call, @NonNull Throwable t) {
                         Log.d("onResponse", t.toString());
+                        if(imageUploadListener !=null){
```
```suggestion
                        if(imageUploadListener != null){
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -1106,17 +1156,17 @@ public void onResponseCallForNotificationPostFunction(Call<State> call, Response
         if (StringUtils.isNotEmpty(imguploadFront)) {
             ProductImage image = new ProductImage(product.getBarcode(), FRONT, new File(imguploadFront));
-            postImg(context, image);
+            postImg(context, image,null);
```
```suggestion
            postImg(context, image, null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -1106,17 +1156,17 @@ public void onResponseCallForNotificationPostFunction(Call<State> call, Response
         String imguploadIngredients = product.getImgupload_ingredients();
         if (StringUtils.isNotEmpty(imguploadIngredients)) {
-            postImg(context, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)));
+            postImg(context, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)),null);
```
```suggestion
            postImg(context, new ProductImage(product.getBarcode(), INGREDIENTS, new File(imguploadIngredients)), null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -1106,17 +1156,17 @@ public void onResponseCallForNotificationPostFunction(Call<State> call, Response
         String imguploadNutrition = product.getImgupload_nutrition();
         if (StringUtils.isNotBlank(imguploadNutrition)) {
-            postImg(context, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)));
+            postImg(context, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)),null);
```
```suggestion
            postImg(context, new ProductImage(product.getBarcode(), NUTRITION, new File(imguploadNutrition)), null);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java:
```diff
@@ -125,6 +130,10 @@
     @BindView(R.id.addPhotoLabel)
     TextView addPhotoLabel;
+    @BindView(R.id.uploading_image_progress)
+    ProgressBar uploading_image_progress;
```
```suggestion
    ProgressBar uploadingImageProgress;
```

brandenfung replied:
Rename to `uploadingImageProgress`

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java:
```diff
@@ -125,6 +130,10 @@
+    @BindView(R.id.uploading_image_progress)
+    ProgressBar uploading_image_progress;
+    @BindView(R.id.uploading_image_progress_text)
+    TextView uploading_image_progress_text;
```
```suggestion
    TextView uploadingImageProgressText;
```

brandenfung replied:
Rename to `uploadingImageProgressText`

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java:
```diff
@@ -708,7 +718,7 @@ public void openFullScreen(View v) {
         ProductImage image = new ProductImage(barcode, FRONT, photoFile);
         image.setFilePath(photoFile.getAbsolutePath());
-        api.postImg(getContext(), image);
+        api.postImg(getContext(), image,this);
```
```suggestion
        api.postImg(getContext(), image ,this);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java:
```diff
@@ -731,7 +741,8 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
                     image.setFilePath(resultUri.getPath());
-                    api.postImg(getContext(), image);
+                    showOtherImageProgress();
+                    api.postImg(getContext(), image,this);
```
```suggestion
                    api.postImg(getContext(), image, this);
```

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/views/product/summary/SummaryProductFragment.java:
```diff
@@ -149,6 +158,7 @@
     private Uri manufactureUri;
     //boolean to determine if image should be loaded or not
     private boolean isLowBatteryMode = false;
+    private Activity activity;
```
Remove `activity`, its not used anymore

brandenfung commented on app/src/main/java/openfoodfacts/github/scrachx/openfood/network/OpenFoodAPIClient.java:
```diff
@@ -336,10 +338,9 @@ public void onFailure(Call<State> call, Throwable t) {
     }
 
-    public void postImg(final Context context, final ProductImage image) {
+    public void postImg(final Context context, final ProductImage image, ImageUploadListener imageUploadListener) {
```
Please refactor to remove duplicated code in `postImg(final Context context, final ProductImage image, ImageUploadListener imageUploadListener)` and `postImg(final Context context, final ProductImage image)`

brandenfung replied:
Everything else looks good, last item here.

jaindiv26 commented:
@brandenfung Please review.

jaindiv26 commented:
@teolemon Can you please re-run Travis build?

teolemon commented:
running

teolemon commented:
Merging, thanks @jaindiv26 and @brandenfung 📸
